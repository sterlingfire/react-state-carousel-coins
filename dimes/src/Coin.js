import "./Coin.css";

const SIDE_TO_URL = {
  Heads: "https://upload.wikimedia.org/wikipedia/commons/d/dd/2017-D_Roosevelt_dime_obverse_transparent.png",
  Tails: "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/2017-D_Roosevelt_dime_reverse_transparent.png/220px-2017-D_Roosevelt_dime_reverse_transparent.png"
} 

/** Renders Coin component
 * 
 * Prop:
 * - Side: "Heads", "Tails" or "" which determines the image
 *   - "" does not display the image
 */

function Coin({side}) {
  let imgUrl = SIDE_TO_URL[side];

  return (<div>
    { side && <img src={imgUrl} className="Coin" alt={side}></img> }
  </div>);
}

export default Coin;
export {SIDE_TO_URL};