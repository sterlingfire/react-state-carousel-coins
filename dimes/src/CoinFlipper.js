import Coin from "./Coin";
import {useState} from 'react';

/* CoinFlipper component
 * props: none
 * state: countHeads:int, countTails:int
 * App -> CoinFlipper -> Coin
 */

function CoinFlipper(){
  const [countHeads, setCountHeads] = useState(0);
  const [countTails, setCountTails] = useState(0);
  const [coinSide, setCoinSide] = useState("");

  function flipCoin(){
    if (Math.random() > 0.5) {
      setCoinSide("Heads");
      setCountHeads(countHeads+1);
    }
    else {
      setCoinSide("Tails");
      setCountTails(countTails+1);
    }
  }

  return (
    <div>
      <h1>Flip A Coin!</h1>
      <Coin side={coinSide}/>
      <button className="CoinFlipper-button" onClick={flipCoin}>Flip Me!</button>
      <p>Out of {countHeads+countTails} flips, there have been {countHeads} heads, and {countTails} tails.</p>
    </div>
  );
}

export default CoinFlipper;
