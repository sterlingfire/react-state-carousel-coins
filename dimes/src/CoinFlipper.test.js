import React from "react";
import { render, fireEvent } from "@testing-library/react";
import CoinFlipper from "./CoinFlipper";
import {SIDE_TO_URL} from "./Coin";

beforeEach(function() {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

afterEach(function() {
  Math.random.mockRestore();
});


it("renders without error", function() {
  render(<CoinFlipper />);
});

it("matches snapshot", function() {
  const { container } = render(<CoinFlipper />);
  expect(container).toMatchSnapshot();
});

it("works when you click on the button", function() {
  const { container, debug } = render(<CoinFlipper />);

  const flipButton = container.querySelector(".CoinFlipper-button");
  fireEvent.click(flipButton);

  const tailsImg = container.querySelector("img");
  debug(container.querySelector("p"));
  const tailsCaption = container.querySelector("p").innerHTML;

  // expect the first coin to be tails
  expect(tailsCaption).toEqual("Out of 1 flips, there have been 0 heads, and 1 tails.u");
  expect(tailsImg.getAttribute("src")).toEqual(SIDE_TO_URL["Tails"]);

  fireEvent.click(flipButton);

  const headsImg = container.querySelector("img");
  const headsCaption = container.querySelector("p").innerHTML;

  // expect the first coin to be tails
  expect(headsCaption).toEqual("Out of 2 flips, there have been 1 heads, and 1 tails.");
  expect(headsImg.getAttribute("src")).toEqual(SIDE_TO_URL["Heads"]);
});
