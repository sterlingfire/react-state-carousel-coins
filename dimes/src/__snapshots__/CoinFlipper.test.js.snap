// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <h1>
      Flip A Coin!
    </h1>
    <div />
    <button
      class="CoinFlipper-button"
    >
      Flip Me!
    </button>
    <p>
      Out of 
      0
       flips, there have been 
      0
       heads, and 
      0
       tails.
    </p>
  </div>
</div>
`;
